Requests by date

kevincasey1222 opened the pull request:
Hi @aiwilliams !

See if this looks good for pulling atSpoke requests by date. I did a little refactoring just to organize the request-pulling termination conditions into one place, and to rename "record" variables to "request" variables.

kevincasey1222 committed:
Refactoring in prep for executionTime aware request pulls

kevincasey1222 committed:
Pulled lastRequestUpdatedAt

kevincasey1222 committed:
Requests now limited by lastUpdated vs execution history

kevincasey1222 committed:
Updated documentation to reflect new behavior

aiwilliams commented:
@kevincasey1222 these commits will need to be rebased on `JupiterOne:master`. I went to pull your branch but could not play the commits on our master.

aiwilliams commented on docs/jupiterone.md:
```diff
@@ -65,11 +65,10 @@ Record entities in the JupiterOne graph, and the collection size will grow
+Consider setting this number to a value a little higher than the number of
+requests typically expected for your **Polling Interval**. If in doubt, go high.
+JupiterOne will stop polling for requests once it reaches any older than the
+last successful date of polling.
```
I think we'll want to drop this configuration variable with these changes. On the first execution, we could go back 14 days.

kevincasey1222 replied:
Did it.

aiwilliams commented on src/steps/requests.ts:
```diff
@@ -14,12 +14,15 @@ import { DATA_ACCOUNT_ENTITY } from './account';
 
   const accountEntity = (await jobState.getData(DATA_ACCOUNT_ENTITY)) as Entity;
+  const lastExecutionTime: number =
+    executionHistory.lastSuccessful?.startedOn || 0;
```
This can be defaulted to a number of days ago.

kevincasey1222 replied:
Defaulted it to 14 days before `new Date()`

aiwilliams commented on src/client.ts:
```diff
@@ -207,38 +213,52 @@ export class APIClient {
+        );
+        if (lastRequestUpdatedAt.getTime() < lastExecutionTime) {
+          lastRequest = true;
+        } //last request pulled is older than the last execution time, so that's enough
```
The logic looks good to me. It would be great to have some tests for this, perhaps starting with pulling the decision out into a function that can be tested without having to simulate API responses? Here are some examples of using Polly.js (the recording stuff) to test the `client.iterateRequests` function with mocked responses.

kevincasey1222 replied:
Alrighty, thorough tests of that function added.

kevincasey1222 committed:
Updated default timeframe for requests to 14 days

kevincasey1222 committed:
Removed unneeded config option NUM_REQUESTS

kevincasey1222 committed:
Pulled out requests loop termination conditions

kevincasey1222 committed:
Tests updated for stopFetchingRequests

kevincasey1222 commented:
Hello @aiwilliams ! Sorry for my slip on not pulling down the latest `JupiterOne: master` before. I've done that now and rebased this branch. Updated the behavior to use a 14 day default, and pulled out the old configuration option NUM_REQUESTS. Manually fixed the recording to match the new URL generated by that change. Pulled out the request fetching termination conditions into a function, and wrote tests to support it.

aiwilliams commented on .env.example:
```diff
@@ -1,2 +1 @@
-CLIENT_ID=
-CLIENT_SECRET=
\ No newline at end of file
+API_KEY=
```
Great catch!

aiwilliams commented on src/steps/index.test.ts:
```diff
@@ -133,3 +133,140 @@ test('should collect data', async () => {
   });
 });
+
+test('stopFetchingRequests', () => {
```
I think it would be good to break these expectations up into separate tests, something like:

```
describe('stopFetchingRequests', () => {
  test('lastExecutionTime undefined ingests max age')
  test('lastExecutionTime 0 ingests max age')
  test('requests missing updateAt ingests limited number')
  test('requests in first page are really old, stops paginating')
  ...
});

kevincasey1222 replied:
Great, thank you, that changed my thought process on tests into something a lot more sophisticated.

aiwilliams commented on src/client.ts:
```diff
@@ -302,22 +307,26 @@ export class APIClient {
-      status: err.status,
-      statusText: "There was a problem parsing the NUM_REQUESTS config field.",
-    });
+export function stopFetchingRequests(requests, lastExecutionTime) {
```
I think the linter should be requiring argument and return typing. Does your environment have the Husky hooks installed in `.git/hooks/pre-commit`?

kevincasey1222 replied:
Yes, they are there. Don't know why it didn't complain on these, but I updated the code with types.

aiwilliams commented on src/steps/index.test.ts:
```diff
@@ -133,3 +133,140 @@ test('should collect data', async () => {
+    updatedAt: '',
+  };
+  requests.push(fakeRequest2);
+  //last request has a falsy updated time, we should keep going
```
It seems that we should probably stop on the first page if there is no other means of stopping. 🤔

kevincasey1222 replied:
Cool. Changed that behavior. Also realized my original test wasn't checking for non-Date()-parsable strings in updatedAt, so adjusted code to handle that properly.

aiwilliams reviewed (changes requested):
Thanks @kevincasey1222 ! My apologies for the long delay in a review.

The `CHANGELOG.md` needs some entries for these changes (the move from number of requests to ingesting up to 14 days ago on the first execution).

kevincasey1222 committed:
Test improvements and some linting

kevincasey1222 commented:
Thanks for the feedback, @aiwilliams ! Tests are much improved now. From your input, I realized a more intelligent way to organize things.

aiwilliams commented on src/steps/index.test.ts:
```diff
@@ -133,3 +137,72 @@ test('should collect data', async () => {
   });
 });
+
+describe('stopFetchingRequests', () => {
```
Thanks for taking up the work to break the tests out. 👍   I'm going to go ahead and merge these changes, but I thought you might like some additional food for thought.

Ideally, each test could be run concurrently and share no state whatsoever. Generally, this means that there would be no variables shared across the `test()` functions, as the `requests`, `lastExecutionTime`, and `templateRequest` are being shared.

1. `lastExecutionTime` - no need for a variable at all, just pass in the value. The variable name is helpful for understanding what the second argument is for the `stopFetchingRequests`, but in the context of a file devoted to testing the function, it is pretty easy to get that from the context.
2. `requests` - just pass in such as `expect(stopFetchingRequests([], 0)).toBe(true);`, or `expect(stopFetchingRequests([buildRequest({ updatedAt: getDaysAgoDateString(13) })], 0)).toBe(true);`. This would drive the development of another test: `test('last request in response is considered oldest')`, where there are two requests with the 0th item have a different timestamp than expected, to be sure it isn't used.
3. `templateRequest` - use a `buildRequest()` function to generate a new value based on defaults (those in the current `templateRequest`). This will avoid mutating a single instance which would be a race condition when executing concurrently.
